flang/lib/Lower/Intrinsics.cpp

```diff
@@ -167,183 +96,168 @@ class IntrinsicLibrary::Implementation {
-    : impl{new Implementation(v, context)} {}
-IntrinsicLibrary::~IntrinsicLibrary() = default;
-
-mlir::Value IntrinsicLibrary::genval(mlir::Location loc,
-                                     Fortran::lower::FirOpBuilder &builder,
-                                     llvm::StringRef name,
-                                     mlir::Type resultType,
-                                     llvm::ArrayRef<mlir::Value> args) const {
-  assert(impl);
-  return impl->genval(loc, builder, name, resultType, args);
+static mlir::FunctionType gen1ArgF32FuncType(mlir::MLIRContext *context) {
+  auto t = mlir::FloatType::getF32(context);
+  return mlir::FunctionType::get({t}, {t}, context);
 }
-
-// MathRuntimeLibrary implementation
-
-// Create the runtime description for the targeted library version.
-// So far ignore the version an only load the dummy llvm lib and pgmath precise
-MathRuntimeLibrary::MathRuntimeLibrary(IntrinsicLibrary::Version,
-                                       mlir::MLIRContext &mlirContext) {
-  for (const MathsRuntimeStaticDescription &func : llvmRuntime) {
-    RuntimeFunction impl{func.getSymbol(),
-                         func.getMLIRFunctionType(&mlirContext)};
-    library.insert({Key{func.getName()}, impl});
-  }
-  for (const MathsRuntimeStaticDescription &func : pgmathPreciseRuntime) {
-    RuntimeFunction impl{func.getSymbol(),
-                         func.getMLIRFunctionType(&mlirContext)};
-    library.insert({Key{func.getName()}, impl});
-  }
+static mlir::FunctionType gen1ArgF64FuncType(mlir::MLIRContext *context) {
+  auto t = mlir::FloatType::getF64(context);
+  return mlir::FunctionType::get({t}, {t}, context);
 }
 
-mlir::FuncOp
-MathRuntimeLibrary::getFuncOp(Fortran::lower::FirOpBuilder &builder,
-                              const RuntimeFunction &runtime) const {
-  auto function = builder.addNamedFunction(runtime.symbol, runtime.type);
-  function.setAttr("fir.runtime", builder.getUnitAttr());
-  return function;
-}
+// TODO : Fill-up this table with more intrinsic.
+// Note: These are also defined as operations in LLVM dialect. See if this
+// can be use and has advantages.
+static constexpr RuntimeFunction llvmIntrinsics[] = {
+    {"abs", "llvm.fabs.f32", gen1ArgF32FuncType},
+    {"abs", "llvm.fabs.f64", gen1ArgF64FuncType},
+    // ceil is used for CEILING but is different, it returns a real.
+    {"ceil", "llvm.ceil.f32", gen1ArgF32FuncType},
+    {"ceil", "llvm.ceil.f64", gen1ArgF64FuncType},
+    {"cos", "llvm.cos.f32", gen1ArgF32FuncType},
+    {"cos", "llvm.cos.f64", gen1ArgF64FuncType},
+    {"log", "llvm.log.f32", gen1ArgF32FuncType},
+    {"log", "llvm.log.f64", gen1ArgF64FuncType},
+    {"log10", "llvm.log10.f32", gen1ArgF32FuncType},
+    {"log10", "llvm.log10.f64", gen1ArgF64FuncType},
+    {"sin", "llvm.sin.f32", gen1ArgF32FuncType},
+    {"sin", "llvm.sin.f64", gen1ArgF64FuncType},
+    {"sqrt", "llvm.sqrt.f32", gen1ArgF32FuncType},
+    {"sqrt", "llvm.sqrt.f64", gen1ArgF64FuncType},
```

Comment:
Ah, right. Still need this mapping.

Reply:
Apart for fabs that I is not in my description of pgmath, we could do all the rest with pgmath runtime.
However, I suspect it might be actually better to throw LLVM intrinsic at LLVM when we can and then to configure LLVM to translate llvm intrinsic to pgmath. So, I have kept all this for further exploration when we are bridged with LLVM.
Maybe Varun has a better view here. I think he taught LLVM to vectorize pgmath functions, I am not sure if LLVM was taught to lower its intrinsics to pgmath.

Reply:
Using the intrinsic is the right thing to do. There can be a mapping in LLVM from the vector forms to pgmath.